File: go/internal/slhdsa/address.go
package slhdsa

import "encoding/binary"

// Address type constants per FIPS 205.
const (
	AddrWotsHash  = 0
	AddrWotsPK    = 1
	AddrTree      = 2
	AddrForsTree  = 3
	AddrForsRoots = 4
	AddrWotsPRF   = 5
	AddrForsPRF   = 6
)

// Address is a 32-byte structure used throughout SLH-DSA.
// Layout (8 uint32 words):
//
//	Word 0:    layer address
//	Word 1-2:  tree address (64-bit, big-endian across words 1-2)
//	Word 3:    type
//	Word 4:    keypair address (or tree index for FORS/XMSS)
//	Word 5:    chain address (or tree height)
//	Word 6:    hash address (or tree index within level)
//	Word 7:    (padding/unused in some contexts)
type Address [32]byte

func (a *Address) setWord(idx int, val uint32) {
	binary.BigEndian.PutUint32(a[idx*4:], val)
}

func (a *Address) getWord(idx int) uint32 {
	return binary.BigEndian.Uint32(a[idx*4:])
}

// SetLayerAddress sets word 0.
func (a *Address) SetLayerAddress(v uint32) {
	a.setWord(0, v)
}

// GetLayerAddress returns word 0.
func (a *Address) GetLayerAddress() uint32 {
	return a.getWord(0)
}

// SetTreeAddress sets the 64-bit tree address (words 1-2).
func (a *Address) SetTreeAddress(v uint64) {
	binary.BigEndian.PutUint64(a[4:12], v)
}

// GetTreeAddress returns the 64-bit tree address.
func (a *Address) GetTreeAddress() uint64 {
	return binary.BigEndian.Uint64(a[4:12])
}

// SetType sets word 3 and zeroes words 4-7.
func (a *Address) SetType(v uint32) {
	a.setWord(3, v)
	// Zero words 4 through 7 on type change
	a.setWord(4, 0)
	a.setWord(5, 0)
	a.setWord(6, 0)
	a.setWord(7, 0)
}

// GetType returns word 3.
func (a *Address) GetType() uint32 {
	return a.getWord(3)
}

// SetKeyPairAddress sets word 4.
func (a *Address) SetKeyPairAddress(v uint32) {
	a.setWord(4, v)
}

// GetKeyPairAddress returns word 4.
func (a *Address) GetKeyPairAddress() uint32 {
	return a.getWord(4)
}

// SetChainAddress sets word 5.
func (a *Address) SetChainAddress(v uint32) {
	a.setWord(5, v)
}

// GetChainAddress returns word 5.
func (a *Address) GetChainAddress() uint32 {
	return a.getWord(5)
}

// SetHashAddress sets word 6.
func (a *Address) SetHashAddress(v uint32) {
	a.setWord(6, v)
}

// GetHashAddress returns word 6.
func (a *Address) GetHashAddress() uint32 {
	return a.getWord(6)
}

// SetTreeHeight sets word 5 (alias for tree contexts).
func (a *Address) SetTreeHeight(v uint32) {
	a.setWord(5, v)
}

// GetTreeHeight returns word 5.
func (a *Address) GetTreeHeight() uint32 {
	return a.getWord(5)
}

// SetTreeIndex sets word 6 (alias for tree contexts).
func (a *Address) SetTreeIndex(v uint32) {
	a.setWord(6, v)
}

// GetTreeIndex returns word 6.
func (a *Address) GetTreeIndex() uint32 {
	return a.getWord(6)
}

// Copy returns a copy of the address.
func (a *Address) Copy() Address {
	var c Address
	copy(c[:], a[:])
	return c
}

// CompressedAddress returns the 22-byte compressed address (ADRSc) used
// in SHA2 variants. It drops the first 3 bytes of each 4-byte word for
// words 0, 3, 4, 5, 6, 7 (keeping byte index 3 of each), and keeps
// 8 bytes for tree address.
func (a *Address) CompressedAddress() []byte {
	// ADRSc layout (22 bytes):
	// byte 0: layer (word 0, byte 3)
	// bytes 1-8: tree address (words 1-2, all 8 bytes)
	// byte 9: type (word 3, byte 3)
	// bytes 10-13: key pair (word 4, all 4 bytes)
	// bytes 14-17: chain/tree height (word 5, all 4 bytes)
	// bytes 18-21: hash/tree index (word 6, all 4 bytes)
	c := make([]byte, 22)
	c[0] = a[3]          // layer: last byte of word 0
	copy(c[1:9], a[4:12]) // tree address: words 1-2
	c[9] = a[15]          // type: last byte of word 3
	copy(c[10:14], a[16:20]) // key pair: word 4
	copy(c[14:18], a[20:24]) // chain/height: word 5
	copy(c[18:22], a[24:28]) // hash/index: word 6
	return c
}

File: swift/Sources/PQCStandards/SlhDsa/Address.swift
// SLH-DSA ADRS (Address) structure

public struct SlhAddress {
    public var data: [UInt8]

    // Address types
    public static let wotsHash: UInt32      = 0
    public static let wotsPk: UInt32        = 1
    public static let tree: UInt32          = 2
    public static let forsTree: UInt32      = 3
    public static let forsRoots: UInt32     = 4
    public static let wotsKeyPrf: UInt32    = 5
    public static let forsKeyPrf: UInt32    = 6

    public init() {
        data = [UInt8](repeating: 0, count: 32)
    }

    // Layer address [0..3]
    public mutating func setLayerAddress(_ v: UInt32) {
        data[0] = UInt8((v >> 24) & 0xFF)
        data[1] = UInt8((v >> 16) & 0xFF)
        data[2] = UInt8((v >> 8) & 0xFF)
        data[3] = UInt8(v & 0xFF)
    }

    // Tree address [4..15]
    public mutating func setTreeAddress(_ v: UInt64) {
        for i in 0..<8 {
            data[4 + (7 - i)] = UInt8((v >> (8 * i)) & 0xFF)
        }
        // Zero the first 4 bytes of tree address for padding
        data[4] = 0; data[5] = 0; data[6] = 0; data[7] = 0
        for i in 0..<8 {
            data[8 + (7 - i)] = UInt8((v >> (8 * i)) & 0xFF)
        }
    }

    public mutating func setTreeAddressBytes(_ v: UInt64) {
        // 12 bytes for tree: data[4..15]
        for i in 0..<4 { data[4 + i] = 0 }
        for i in 0..<8 {
            data[8 + (7 - i)] = UInt8((v >> (8 * i)) & 0xFF)
        }
    }

    // Type [16..19]
    public mutating func setType(_ v: UInt32) {
        data[16] = UInt8((v >> 24) & 0xFF)
        data[17] = UInt8((v >> 16) & 0xFF)
        data[18] = UInt8((v >> 8) & 0xFF)
        data[19] = UInt8(v & 0xFF)
        // Zero padding after type
        for i in 20..<32 { data[i] = 0 }
    }

    // Keypair address [20..23]
    public mutating func setKeyPairAddress(_ v: UInt32) {
        data[20] = UInt8((v >> 24) & 0xFF)
        data[21] = UInt8((v >> 16) & 0xFF)
        data[22] = UInt8((v >> 8) & 0xFF)
        data[23] = UInt8(v & 0xFF)
    }

    // Chain address [24..27]
    public mutating func setChainAddress(_ v: UInt32) {
        data[24] = UInt8((v >> 24) & 0xFF)
        data[25] = UInt8((v >> 16) & 0xFF)
        data[26] = UInt8((v >> 8) & 0xFF)
        data[27] = UInt8(v & 0xFF)
    }

    // Hash address [28..31]
    public mutating func setHashAddress(_ v: UInt32) {
        data[28] = UInt8((v >> 24) & 0xFF)
        data[29] = UInt8((v >> 16) & 0xFF)
        data[30] = UInt8((v >> 8) & 0xFF)
        data[31] = UInt8(v & 0xFF)
    }

    // Tree height [24..27]
    public mutating func setTreeHeight(_ v: UInt32) {
        setChainAddress(v)
    }

    // Tree index [28..31]
    public mutating func setTreeIndex(_ v: UInt32) {
        setHashAddress(v)
    }

    public func copy() -> SlhAddress {
        var a = SlhAddress()
        a.data = self.data
        return a
    }
}
